src/S3/S3Transfer/AbstractMultipartUploader.php

```diff
@@ -0,0 +1,432 @@
+     * @param array $config
+     * @param string|null $uploadId
+     * @param array $parts
+     * @param TransferProgressSnapshot|null $currentSnapshot
+     * @param TransferListenerNotifier|null $listenerNotifier
+     */
+    public function __construct
+    (
+        S3ClientInterface $s3Client,
+        array $createMultipartArgs,
+        array $config,
+        ?string $uploadId = null,
+        array $parts = [],
+        ?TransferProgressSnapshot $currentSnapshot = null,
+        ?TransferListenerNotifier $listenerNotifier = null,
+    )
+    {
+        $this->s3Client = $s3Client;
+        $this->createMultipartArgs = $createMultipartArgs;
+        $this->validateConfig($config);
+        $this->config = $config;
+        $this->uploadId = $uploadId;
+        $this->parts = $parts;
+        $this->currentSnapshot = $currentSnapshot;
+        $this->listenerNotifier = $listenerNotifier;
+    }
+
+    /**
+     * @param array $config
+     *
+     * @return void
+     */
+    protected function validateConfig(array &$config): void
+    {
+        if (!empty($partSize = $config['part_size'])
+            && ($partSize < self::PART_MIN_SIZE || $partSize > self::PART_MAX_SIZE)
+        ) {
+                throw new \InvalidArgumentException(
+                    "The config `part_size` value must be between "
+                    . self::PART_MIN_SIZE . " and " . self::PART_MAX_SIZE . "."
+                    . self::PART_MIN_SIZE . " and " . self::PART_MAX_SIZE
+                    . " but ${config['part_size']} given."
+                );
+            }
+
+        $config['part_size'] = self::PART_MIN_SIZE;
+    }
+
+    /**
+     * @return string|null
+     */
+    public function getUploadId(): ?string
+    {
+        return $this->uploadId;
+    }
+
+    /**
+     * @return array
+     */
+    public function getParts(): array
+    {
+        return $this->parts;
+    }
+
+    /**
+     * Get the current progress snapshot.
+     * @return TransferProgressSnapshot|null
+     */
+    public function getCurrentSnapshot(): ?TransferProgressSnapshot
+    {
+        return $this->currentSnapshot;
+    }
+    /**
+     * @return PromiseInterface
+     */
+    public function promise(): PromiseInterface
+    {
+        return Coroutine::of(function () {
+            try {
+                yield $this->createMultipartUpload();
+                yield $this->processMultipartOperation();
+                $result = yield $this->completeMultipartUpload();
+                yield Create::promiseFor($this->createResponse($result));
+            } catch (Throwable $e) {
+                $this->operationFailed($e);
+                yield Create::rejectionFor($e);
+            } finally {
+                $this->callDeferredFns();
+            }
+        });
+    }
+
+    /**
+     * @return PromiseInterface
+     */
+    protected function createMultipartUpload(): PromiseInterface
+    {
+        $requestArgs = [...$this->createMultipartArgs];
```

Comment:
This is probably logic borrowed from https://github.com/aws/aws-sdk-php/pull/3079, but because `createMultipartArgs` is already an array and we're not combining it with another array, I think we just just pass `$this->createMultipartArgs` to `operationInitiated()`